FreeRADIUS/freeradius-server
# Add native support for Redis 3.0 cluster

Overview
=======

 Read and understand this http://redis.io/topics/cluster-spec first, else the text below
 will not be useful.

Testing
---------

```
./scripts/travis/redis-setup.sh
export REDIS_TEST_SERVER=127.0.0.1
make test
```

More tests needed in future, but included tests are sufficient to verify basic functionality, and capability to handle cluster failover events.


Using the cluster code's public API
----------------------------------------------

 The two functions used at runtime to issue commands are #fr_redis_cluster_state_init
 and #fr_redis_cluster_state_next.

 #fr_redis_cluster_state_init initialises the structure we use to track which node
 we're communicating with, and uses a key (string) to determine which node we should
 try and contact first.

 #fr_redis_cluster_state_next examines the result of the last command, and either
 gets a new connection, or errors out.

 In both cases the connection should not be released by the caller, only by
 #fr_redis_cluster_state_next.

 Below is the sequence of calls required to use the cluster:

     1. During initialization allocate a cluster configuration structure with
	  #fr_redis_cluster_alloc.  This holds the cluster configuration and state.
     2. Declare a #fr_redis_cluster_state_t in the function that needs to issue
        commands against the cluster.
     3. Call #fr_redis_cluster_state_init with relevant arguments including a pointer
        to the #fr_redis_cluster_state_t struct, and a key/key_len used for initial
        lookup.  For most commands the key is the value of the second argument.
        #fr_redis_cluster_state_init will then and reserve/pass back a connection for
        the pool associated with the node associated with the key.
     4. Use the connection that was passed back, to issue a Redis command against.
     5. Use #fr_redis_command_status to translate the result of the command into
        a #redis_rcode_t value.
     6. Call #fr_redis_cluster_state_next with relevant arguments including a pointer
        to the #fr_redis_cluster_state_t struct and the #redis_rcode_t value for the
        last command.
     7. If #fr_redis_cluster_state_next returns 0, repeat steps 4-7.  Otherwise
        stop and analyse the return value.

 See #fr_redis_cluster_state_init for example code.

Structures
========

   This code maintains a series structures for efficient lookup and lockless operations.

   The important ones are:
     - An array of #cluster_node_t.  These are pre-allocated on startup and are
       never added to, or removed from.
     - An #fr_fifo_t.  This contains the queue of nodes that may be re-used.
     - An #rbtree_t.  This contains a tree of nodes which are active.  The tree is built on IP
       address and port.

   Each #cluster_node_t contains a master ID, and an array of slave IDs.  The IDs are array
   indexes in the fr_redis_cluster_t.node array.  We use 8bit unsigned integers instead of
   pointers to save space.  Using pointers, the node[] array would need 784K, using IDs
   it uses 112K.  Still not light on memory, but a bit more acceptable.
   Currently the key_slot array is shadowed by key_slot_pending, used to stage new key_slot
   mappings.  This doubles the memory used.  We may want to consider allocating key_slot_pending
   only during remappings and freeing it after.

Mapping/Remapping the cluster
========================

   On startup, and during cluster operation, a remap may be performed.  A remap involves
   the following steps:

     1. Executing the Redis 'cluster slots' command.
     2. Validating the result of this command.  We need to do extensive validation to
        avoid SEGV on invalid data, due to the way libhiredis presents the result.
     3. Determining the intersection between nodes described in the result, and those already
        in our #rbtree_t.
     4. Connecting to nodes that were in the result, but not in the tree.
        Note: If we can't connect to any of the masters, we count the map as invalid, roll
        back any newly connected nodes, and error out. Slave failure is OK.
     5. Mapping keyslot ranges to nodes in the key_slot_pending array.
     6. Verifying there are no holes in the ranges (if there are, we roll back and error out).
     7. Applying the new keyslot ranges.
     8. Removing nodes no longer used by the key slots, and adding them back to the free
        nodes queue.

   #cluster_map_get and #cluster_map_apply, perform the operations described
   above. The get function, issues the 'cluster slots' command and performs validation, the
   apply function processes and applys the map.

   Failing to apply a map is not a fatal error at runtime, and is only fatal on startup if
   pool.start > 0.

   The cluster client can continue to operate, albeit inefficiently, with a stale cluster map
   by following '-ASK' and '-MOVE' redirects.

   Remaps are limited to one per second.  If any operation sets the remap_needed flag, or
   attempts a remap directly, the remap may be skipped if one occurred recently.


Processing '-ASK' and '-MOVE' redirects
===============================


   The code treats '-ASK' (temporary redirect) and '-MOVE' (permanent redirect) responses
   similarly.  If the node is known, then a connection is reserved from its pool, if the node
   is not known, a new pool is established, and a connection reserved.

   The difference between '-ASK' and '-MOVE' is that '-MOVE' attempts a cluster remap before
   following the redirect.

   The data from '-MOVE' responses, is not used to alter the cluster map.  That is only done
   on successful remap.


Processing '-TRYAGAIN'
===================

   If the cluster is in a state of flux, a node may return '-TRYAGAIN' to indicated that we
   should attempt the operation again.  The cluster spec says we should attempt the operation
   after some time.  This time is configurable.
